chicago/permit_cleaning.py

```diff
@@ -70,31 +70,55 @@ def download_permits(start_date, end_date):
+    This function creates new rows for each PIN in multi-PIN permits and saves the relevant PIN in pin_solo.
     """
-    # the downloaded dataframe will not include any pin columns that are completely blank, so check for existing ones here
-    all_pin_columns = ["pin1", "pin2", "pin3", "pin4", "pin5", "pin6", "pin7", "pin8", "pin9", "pin10"]
-    pin_columns = [col for col in df.columns if col in all_pin_columns]
-    non_pin_columns = [col for col in df.columns if col not in pin_columns]
+    def parse_pin_list(pin_list_str):
+        """Parse the PIN_LIST column. The column is formatted like so:
 
-    melted_df = pd.melt(df, id_vars=non_pin_columns, value_vars=pin_columns, var_name="pin_type", value_name="solo_pin")
+          * If no PINs are in the list, the value is NA, and we want to keep it
+          * Otherwise, PINs are stored as pipe-separated values, and we want to
+            parse them as lists so that we can pivot them out to one
+            permit per PIN
+        """
+        if pd.isna(pin_list_str):
+            return np.nan
+        elif " | " in pin_list_str:
+            pin_list = pin_list_str.split(" | ")
+            # Remove duplicates while maintaining list order
+            return list(dict.fromkeys(pin_list))
+        else:
+            return [pin_list_str]
+
+    df['pin_list'] = df['pin_list'].apply(parse_pin_list)
 
-    # keep rows with NA for pin1, filter out rows with NA for other pins
-    melted_df = melted_df[(melted_df["pin_type"] == "pin1") | ((melted_df["pin_type"] != "pin1") & melted_df["solo_pin"].notna())]
+    # Retain rows where pin_list is NA, since the pivot operation will
+    # remove them but we want to keep them
+    na_rows = df[df["pin_list"].isna()]
+    df = df.dropna(subset=["pin_list"])
 
-    # Sometimes the incoming permits have duplicate PINs across PIN
-    # columns, e.g. pin3 == pin4 and both are not null. Make sure to catch
-    # this edge case and remove the duplicate PINs from the set, giving
-    # priority to PINs with lower pin_types (e.g. if pin2 and pin3 share the
-    # same PIN, pin2 should take priority)
-    melted_df = melted_df.sort_values(by=["permit_", "solo_pin", "pin_type"]).drop_duplicates(subset=["permit_", "solo_pin"], keep="first")
+    # Pivot the dataframe longer by "pin_list" so that we have one row
+    # per PIN with "solo_pin" being the PIN column
+    df = df.explode("pin_list").reset_index(drop=True).rename(columns={
+        "pin_list": "solo_pin"
+    })
 
-    # order rows by permit number then pin type (so pins will be in order of their assigned numbering in permit table, not necessarily by pin number)
-    melted_df = melted_df.sort_values(by=["permit_", "pin_type"]).reset_index(drop=True)
+    # Add a column to track the position of the PIN in the PIN list for
+    # ordering, starting at 1
+    df["pin_type"] = df.groupby(level=0).cumcount() + 1
```

Comment:
**question (blocking):** Is this the right grouping? The index is reset in the line above, so I think this will return the global row count, rather than by PIN.

Reply:
Great catch, I switched this up to group by `permit_` in 0095269b9651dbd2a02e8292a4b1b5c554a8d648 and spot checked two permits with multiple PINs to confirm the `pin_type` column was set correctly 👍🏻